stl/inc/xstring

```diff
@@ -2432,6 +2440,106 @@ public:
+        _STL_INTERNAL_CHECK(_Right_size <= max_size());
+        _STL_INTERNAL_CHECK(_Right_size <= max_size() - _Left_size);
+        const auto _New_size    = static_cast<size_type>(_Left_size + _Right_size);
+        size_type _New_capacity = _BUF_SIZE - 1;
+        auto& _My_data          = _Mypair._Myval2;
+        _Elem* _Ptr             = _My_data._Bx._Buf;
+        auto&& _Alproxy         = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
+        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _My_data); // throws
+        if (_New_capacity < _New_size) {
+            _New_capacity           = _Calculate_growth(_New_size, _BUF_SIZE - 1, max_size());
+            const pointer _Fancyptr = _Getal().allocate(_New_capacity + 1); // throws
+            _Ptr                    = _Unfancy(_Fancyptr);
+            _Construct_in_place(_My_data._Bx._Ptr, _Fancyptr);
+        }
+
+        _My_data._Mysize = _New_size;
+        _My_data._Myres  = _New_capacity;
+        _Traits::copy(_Ptr, _Left_ptr, _Left_size);
+        _Traits::copy(_Ptr + static_cast<ptrdiff_t>(_Left_size), _Right_ptr, _Right_size);
+        _Traits::assign(_Ptr[_New_size], _Elem());
+        _Proxy._Release();
+    }
+
+    basic_string(_String_constructor_concat_tag, basic_string& _Left, basic_string& _Right)
+        : _Mypair(_One_then_variadic_args_t(), _Left._Getal()) {
+        auto& _My_data    = _Mypair._Myval2;
+        auto& _Left_data  = _Left._Mypair._Myval2;
+        auto& _Right_data = _Right._Mypair._Myval2;
+        _Left_data._Orphan_all();
+        _Right_data._Orphan_all();
+        const auto _Left_size  = _Left_data._Mysize;
+        const auto _Right_size = _Right_data._Mysize;
+
+        const auto _Left_capacity  = _Left_data._Myres;
+        const auto _Right_capacity = _Right_data._Myres;
+        // overflow is OK due to max_size() checks:
+        const auto _New_size = static_cast<size_type>(_Left_size + _Right_size);
+        if (_Right_size <= _Left_capacity - _Left_size && _Left_capacity >= _Right_capacity) {
+            // take _Left's buffer, max_size() is OK because the result fits in _Left's capacity
+            _My_data._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal())); // throws, hereafter nothrow in this block
+            _Take_contents(_Left, bool_constant<_Can_memcpy_val>{});
+            const auto _Ptr = _My_data._Myptr();
+            _Traits::copy(_Ptr + _My_data._Mysize, _Right_data._Myptr(), _Right_size);
+            _My_data._Mysize = _New_size;
+            _Traits::assign(_Ptr[_New_size], _Elem());
+            return;
+        }
+
+        if (_Allocators_equal(_Getal(), _Right._Getal()) && _Left_size <= _Right_capacity - _Right_size) {
+            // take _Right's buffer, max_size() is OK because the result fits in _Right's capacity
+            // At this point, we have tested:
+            // !(_Right_size <= _Left_capacity - _Left_size && _Left_capacity >= _Right_capacity)

```

Comment:
These expressions would be simpler if you introduced bools named like `_Right_fits_within_Left` or `_Left_has_room_for_Right`. Or better yet, you might be able to name only one side, like `_Fits_in_Left` or `_Left_has_space` or something. The highly repetitive left right left right size capacity size stuff makes this far more difficult to read than it should be.

Reply:
I was not optimistic that the codegen for this would be OK: optimizers tend to get confused by switching between the value domain and the control flow domain like this. However, I tested and the output matches, so I applied this.